Completed chapters 0-6

bl3nd3rm4n opened the pull request.

bl3nd3rm4n committed:
Completed chapters 0-6

Ducuvlad commented on online-shop/src/app/components/product-details/product-details.component.html:
```diff
@@ -0,0 +1,26 @@
+    <button (click)="edit()" *ngIf="authService.isAdmin()" class="default-border" id="edit-button">Edit</button>
+    <button (click)="delete()" *ngIf="authService.isAdmin()" class="default-border" id="delete-button">Delete</button>
+    <div class="default-border product-information">
+      <div class="divTable properties">
```
Please use \<table\> and the associated table tags (tr, td , th) instead of \<div\>

bl3nd3rm4n committed:
Made the tables out of <table> and the associated table tags (tr, td , th) instead of <div>

flyforfun98 commented on online-shop/src/app/components/nav-bar/nav-bar.component.css:
```diff
@@ -0,0 +1,28 @@
+}
+
+.nav-separator {
+  width: 4px;;
```
You typed here twice the ';'. Watch out at this kind of mistakes cause the build could fail.

flyforfun98 commented on online-shop/src/app/components/product-editor/product-editor.component.html:
```diff
@@ -0,0 +1,76 @@
+<div *ngIf="product || isAddMode" class="product-details">
+  <div class="details-container">
+    <h1 *ngIf="!isAddMode">Edit Product: {{originalProduct?.name}}</h1>
+    <h1 *ngIf="isAddMode">New Product</h1>
```
Practically this is the else branch of line 3 in this file. You could use the template ***ngIf = "condition; else template"**.
See this link https://stackoverflow.com/questions/43006550/how-to-use-ngif-else

flyforfun98 commented on online-shop/src/app/components/product-editor/product-editor.component.ts:
```diff
@@ -0,0 +1,59 @@
+  }
+
+  getProduct(): void {
+    const id = Number(this.route.snapshot.paramMap.get('id'));
```
You could pass the id of the product as a parameter to not duplicate the code
getProduct(productId: string)

flyforfun98 commented on online-shop/src/app/components/product-editor/product-editor.component.ts:
```diff
@@ -0,0 +1,59 @@
+    const id = Number(this.route.snapshot.paramMap.get('id'));
+    this.productService.getProduct(id)
+      .subscribe(product => {
+          this.originalProduct = JSON.parse(JSON.stringify(product));
```
Do you really need JSON.parse(JSON.stringify... here?
Can't you just assign to originalProduct the value of product?

bl3nd3rm4n replied:
No, because then the name of the product in the title (Edit Product: ...) will change as the user edits it.
I could save just the name of the product in a variable if you'd like it better that way.

flyforfun98 commented on online-shop/src/app/components/product-list/product-list.component.css:
```diff
@@ -0,0 +1,21 @@
+  justify-content: center;
+}
+
+.product-list-container {
```
This css class be deleted since it is empty.

flyforfun98 commented on online-shop/src/app/components/product-list/product-list.component.html:
```diff
@@ -0,0 +1,19 @@
+<div class="product-list">
+  <div class="product-list-container">
```
product-list-container can be removed since the css class corresponding to it is empty.

flyforfun98 commented on online-shop/src/app/components/shopping-cart/shopping-cart.component.css:
```diff
@@ -0,0 +1,25 @@
+  justify-content: center;
+}
+
+.shopping-cart-container {
```
This can be removed too.

Ducuvlad replied:
Please remove unused class.

flyforfun98 commented on online-shop/src/app/components/shopping-cart/shopping-cart.component.html:
```diff
@@ -0,0 +1,24 @@
+<div class="shopping-cart">
+  <div class="shopping-cart-container">
```
shopping-cart-container can be removed too.

flyforfun98 commented on online-shop/src/app/services/auth.service.ts:
```diff
@@ -0,0 +1,62 @@
+  providedIn: 'root'
+})
+export class AuthService {
+  user!: User;
```
You don't have to use '!' to every object. For example in this case you can assign null to the 'user' variable when the user is logged out.

bl3nd3rm4n replied:
TS2322: Type 'null' is not assignable to type 'User'.

flyforfun98 commented on online-shop/src/app/services/auth.service.ts:
```diff
@@ -0,0 +1,62 @@
+  }
+
+  hasRole(role: Role): boolean {
+    return this.user?.roles.includes(role);
```
'?' is used to indicate the optionality. However it is recommended to use it in HTML or as a parameter in the method.
Thus, you could rewrite your code as 



     hasRole(role: Role, user?: User): boolean {   
             return user.roles.includes(role); 
     }

Or you could use If - else clause to handle the logic to not pass 'user' as a parameter.

bl3nd3rm4n replied:
Besides Internet Explorer support I really don't see any point in having the user as a parameter, nor in handling the logic through if-else. Is that your concern?

flyforfun98 commented on online-shop/src/app/services/cart.service.ts:
```diff
@@ -0,0 +1,62 @@
+  add(id: number): Observable<Map<number, number>> {
+    this.productIdQuantityMap.set(id, this.productIdQuantityMap.has(id) ? this.productIdQuantityMap.get(id)! + 1 : 1);
+    this.saveToLocalStorage();
+    return of(this.productIdQuantityMap);
```
Why are you returning the map if you don't use its value? I have seen this 'add' method being called just in product-details.component.ts and method addToCart from there does not return anything. The same issue happens with method updateQuantity (line 32).

bl3nd3rm4n replied:
The idea was to have the same return type as the other similar methods (getProducts, remove and checkout).
Will make them void

bl3nd3rm4n committed:
Fixed some comments on PR
- removed unused classes
- used ng-templates for if-else logic in product-editor.component.html
- added id as a parameter for the getProduct method in product-editor.component.ts
- removed unused return values in cart.service.ts
